modules/delegates.js

```diff
@@ -353,112 +353,160 @@ __private.decryptPassphrase = function(encryptedPassphrase, password) {
  * @returns {setImmediateCallback} cb, err
  * @todo Add description for the params
  */
-__private.checkDelegates = function(publicKey, votes, state, cb, tx) {
+__private.checkDelegates = function(senderPublicKey, votes, state, cb, tx) {
 	if (!Array.isArray(votes)) {
 		return setImmediate(cb, 'Votes must be an array');
 	}
 
-	modules.accounts.getAccount(
-		{ publicKey },
-		(err, account) => {
-			if (err) {
-				return setImmediate(cb, err);
-			}
+	// Converting publicKeys to addresses because module.accounts do not support fetching multiple accounts by publicKeys
+	// TODO: Use Hashmap to improve performance further.
+	const actionWithAddresses = votes.map(vote => {
+		const action = vote[0];
+		let voteAddress;
 
-			if (!account) {
-				return setImmediate(cb, 'Account not found');
+		try {
+			if (action !== '+' && action !== '-') {
+				throw 'Invalid math operator';
 			}
+			const votePublicKey = vote.slice(1);
 
-			const delegates =
-				state === 'confirmed' ? account.delegates : account.u_delegates;
-			const existingVotes = Array.isArray(delegates) ? delegates.length : 0;
-			let additions = 0;
-			let removals = 0;
-
-			async.eachSeries(
-				votes,
-				(action, cb) => {
-					const math = action[0];
-
-					if (math === '+') {
-						additions += 1;
-					} else if (math === '-') {
-						removals += 1;
-					} else {
-						return setImmediate(cb, 'Invalid math operator');
-					}
-
-					const publicKey = action.slice(1);
-
-					try {
-						Buffer.from(publicKey, 'hex');
-					} catch (e) {
-						library.logger.error(e.stack);
-						return setImmediate(cb, 'Invalid public key');
-					}
-
-					modules.accounts.getAccount(
-						{ publicKey, isDelegate: 1 },
-						(err, account) => {
-							if (err) {
-								return setImmediate(cb, err);
-							}
-
-							if (!account) {
-								return setImmediate(cb, 'Delegate not found');
-							}
-
-							if (
-								math === '+' &&
-								(delegates != null && delegates.indexOf(publicKey) !== -1)
-							) {
-								return setImmediate(
-									cb,
-									`Failed to add vote, delegate "${
-										account.username
-									}" already voted for`
-								);
-							}
-
-							if (
-								math === '-' &&
-								(delegates === null || delegates.indexOf(publicKey) === -1)
-							) {
-								return setImmediate(
-									cb,
-									`Failed to remove vote, delegate "${
-										account.username
-									}" was not voted for`
-								);
-							}
-
-							return setImmediate(cb);
-						},
-						tx
-					);
-				},
-				err => {
-					if (err) {
-						return setImmediate(cb, err);
-					}
-
-					const totalVotes = existingVotes + additions - removals;
+			voteAddress = modules.accounts.generateAddressByPublicKey(votePublicKey);
+		} catch (e) {
+			library.logger.error(e.stack);
+			return setImmediate(cb, e);
```

Comment:
Does this `return` will stop the consequent iterations of `votes.map` on line 363 ?

Reply:
Yes. Since javascript is single threaded and there are no asynchronous calls in this `map`. It will not continue any further. But I get your point that it makes the code look more imperative. Do you have any suggestions? I could have one `.each` for verifying and another `map` for converting votes to actionsWithAddresses but that seemed less performant.

Reply:
I suggest to move `try/catch` outside `map` and use the callback there.